File: frontend/src/app/models/token.ts
import { Ability } from './ability';

export interface CharacterSheet {
  class: string;
  level: number;
  background: string;
  playerName: string;
  race: string;
  alignment: string;
  xp: number;
  hitDie: number;
  
  str: number;
  dex: number;
  con: number;
  int: number;
  wis: number;
  cha: number;
  
  ac: number;
  initiative: number;
  speed: number;
  
  proficiencyBonus: number;
  passivePerception: number;
  
  hp: number;
  maxHp: number;
  spellUses?: number;
  maxSpellUses?: number;
  spellcastingAbility?: 'int' | 'wis' | 'cha';
  
  cp?: number;
  sp?: number;
  ep?: number;
  gp?: number;
  pp?: number;
  backpack?: string;
  inventory?: { name: string; quantity: number; weight: number; isEquipped: boolean }[];
  exhaustion?: number;
  restLevel?: number;
  rations?: number;
  water?: number;
  savingThrowProficiencies?: string[];
  skillProficiencies?: string[];
  expertiseSkills?: string[];
  
  /** Resistências a tipos de dano (PHB p.197) — dano reduzido à metade */
  damageResistances?: string[];
  /** Imunidades a tipos de dano — dano = 0 */
  damageImmunities?: string[];
  /** Vulnerabilidades a tipos de dano — dano dobrado */
  damageVulnerabilities?: string[];
}

export interface TokenCondition {
  id: string;
  name: string;
  icon: string;
  color: string;
  damagePerTurn?: number;
}

export interface Token {
  id: string;
  name: string;
  characterId?: string;
  x: number;
  y: number;
  hp: number;
  maxHp: number;
  spellUses?: number;
  maxSpellUses?: number;
  conditions: TokenCondition[];
  controlledBy: string;
  color: string;
  imageUrl?: string;
  imageScale?: number;
  imageOffsetX?: number;
  imageOffsetY?: number;
  type?: 'player' | 'enemy' | 'npc' | 'boss' | 'item';
  abilities?: Ability[];
  sheet?: CharacterSheet;
  initiative?: number;
  xpReward?: number; // XP concedido aos jogadores ao derrotar este inimigo (baseado no ND)
}

File: backend/src/main/java/com/blooddragons/model/Token.java
package com.blooddragons.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

/**
 * Token no mapa de combate (personagem, inimigo, NPC, boss ou item).
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Token {
    private String id;
    private String name;
    private String characterId;
    private int x;
    private int y;
    private int hp;
    private int maxHp;
    private Integer spellUses;
    private Integer maxSpellUses;

    @Builder.Default
    private List<TokenCondition> conditions = new ArrayList<>();

    private String controlledBy;
    private String color;
    private String imageUrl;
    private Double imageScale;
    private Double imageOffsetX;
    private Double imageOffsetY;

    /** player | enemy | npc | boss | item */
    private String type;

    @Builder.Default
    private List<Ability> abilities = new ArrayList<>();

    private CharacterSheet sheet;
    private Integer initiative;
}
